crud/src/main/java/com/redhat/lightblue/mediator/Mediator.java

```diff
@@ -662,30 +663,47 @@ protected void wait(BulkExecutionContext ctx) {
         }
     }
 
-    protected Callable<Response> getFutureRequest(final Request req) {
+    protected Callable<Response> getFutureRequest(final Request req, RequestMetrics metrics) {
         return new Callable<Response>() {
             @Override
             public Response call() {
                 LOGGER.debug("Starting a future {} request",req.getOperation());
+                RequestMetrics.Context metricCtx = null;
+                Response resp = null;
                 switch (req.getOperation()) {
                     case FIND:
-                        return find((FindRequest) req);
+                        metricCtx = metrics.startEntityRequest("find", req.getEntityVersion().getEntity(), req.getEntityVersion().getVersion());
+                        resp = find((FindRequest) req);
+                        metricCtx.markAllErrorsAndEndRequestMonitoring(resp.getErrors());
+                        return resp;
                     case INSERT:
-                        return insert((InsertionRequest) req);
+                        metricCtx = metrics.startEntityRequest("insert", req.getEntityVersion().getEntity(), req.getEntityVersion().getVersion());
+                        resp = insert((InsertionRequest) req);
+                        metricCtx.markAllErrorsAndEndRequestMonitoring(resp.getErrors());
+                        return resp;
                     case DELETE:
-                        return delete((DeleteRequest) req);
+                        metricCtx  = metrics.startEntityRequest("delete", req.getEntityVersion().getEntity(), req.getEntityVersion().getVersion());
+                        resp = delete((DeleteRequest) req);
+                        metricCtx.markAllErrorsAndEndRequestMonitoring(resp.getErrors());
+                        return resp; 
                     case UPDATE:
-                        return update((UpdateRequest) req);
+                        metricCtx = metrics.startEntityRequest("update", req.getEntityVersion().getEntity(), req.getEntityVersion().getVersion());
+                        resp = update((UpdateRequest) req);
+                        metricCtx.markAllErrorsAndEndRequestMonitoring(resp.getErrors());
+                        return resp;
                     case SAVE:
-                        return save((SaveRequest) req);
+                        metricCtx = metrics.startEntityRequest("save", req.getEntityVersion().getEntity(), req.getEntityVersion().getVersion());
+                        resp = save((SaveRequest) req);
+                        metricCtx.markAllErrorsAndEndRequestMonitoring(resp.getErrors());
+                        return resp;
```

Comment:
This looks good. For what it's worth, you could probably refactor so that we store the response in the switch, then outside of the switch we have a single `metricCtx.markAllErrorsAndEndRequestMonitoring(resp.getErrors());` and a single return. It reduces duplication a little bit. Not a huge deal though.

Also, we could consider adding an overload to `startEntityRequest` that accepts a `CRUDOperation`. Then we could have a single `startEntityRequest` call as well.